scripts/src/java/org/oppia/android/scripts/todo/TodoCheck.kt

```diff
@@ -0,0 +1,302 @@
+    failureMessage = "TODOs not in correct format:",
+    failureNote =
+      "The TODO should be in the format: TODO(#ISSUE_NUMBER): <todo_description>"
+  )
+
+  logFailures(
+    invalidTodos = openIssueFailureTodosAfterExemption,
+    failureMessage = "TODOs not corresponding to an open issue:",
+    failureNote =
+      "Note that, every TODO must correspond to an open issue on GitHub"
+  )
+
+  if (
+    redundantExemptions.isNotEmpty() ||
+    poorlyFormattedTodosAfterExemption.isNotEmpty() ||
+    openIssueFailureTodosAfterExemption.isNotEmpty()
+  ) {
+    throw Exception("TODO CHECK FAILED")
+  } else {
+    println("TODO CHECK PASSED")
+  }
+}
+
+/**
+ * Retrieves the todo failures list after filtering them from the exemptions.
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return list obtained after filtering the exemptions
+ */
+private fun retrieveTodosAfterExemption(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Todo> {
+  return todos.filter { todo ->
+    !todoExemptionList.any { it ->
+      it.exemptedFilePath == todo.filePath.removePrefix(repoPath) &&
+        todo.lineNumber in it.getLineNumberList()
+    }
+  }
+}
+
+/**
+ * Retrieves a list of redundant exemptions.
+ *
+ * @param todos the list of all the failure causing todos
+ * @param todoExemptionList the list contating the todo exemptions
+ * @param repoPath path of the repo to be analyzed
+ * @return a list of all the redundant exemptions
+ */
+private fun retrieveRedundantExemptions(
+  todos: List<Todo>,
+  todoExemptionList: List<TodoExemption>,
+  repoPath: String
+): List<Pair<String, Int>> {
+  return todoExemptionList.flatMap { exemption ->
+    exemption.getLineNumberList().mapNotNull { exemptedLineNumber ->
+      if (
+        !todos.any {
+          it.filePath.removePrefix(repoPath) == exemption.exemptedFilePath &&
+            it.lineNumber == exemptedLineNumber
+        }
+      ) {
+        Pair(exemption.exemptedFilePath, exemptedLineNumber)
+      } else {
+        null
+      }
+    }
+  }
+}
+
+/**
+ * Checks whether a line of code contains a poorly formatted todo.
+ *
+ * @param codeLine the line of code to be checked
+ * @param todoRegex regex of todo
+ * @param todoStartingRegex regex for the starting of the todo
+ * @param correctTodoFormatRegex regex of the correct todo format
+ * @return whether the line contains a poorly formatted TODO
+ */
+private fun checkIfTodoIsPoorlyFormatted(
+  codeLine: String,
+  todoRegex: Regex,
+  todoStartingRegex: Regex,
+  correctTodoFormatRegex: Regex
+): Boolean {
+  if (todoStartingRegex.containsMatchIn(codeLine)) {
+    if (!correctTodoFormatRegex.containsMatchIn(codeLine)) {
+      return true
+    }
+    return false
+  }
+  if (todoRegex.containsMatchIn(codeLine)) {
+    return true
+  }
+  return false
```

Comment:
Are each of these possible ``return`` outcomes covered in tests?

Reply:
Yes, each case is well covered in the tests.